physics/sfc_diag.f

```diff
@@ -75,23 +95,131 @@ subroutine sfc_diag_run                                           &
 !       t2m(i)  = t2m(i) * sig2k
         wrk     = 1.0 - fhi
 
+        thcon    = (1.e5/ps(i))**rocp    
+       !-- make sure 1st level q is not higher than saturated value
+        qss    = fpvs(t1(i))
+        qss    = eps * qss / (ps(i) + epsm1 * qss)
+        q1c    = min(q1(i),qss) ! lev 1 spec. humidity
 
-        if(thsfc_loc) then ! Use local potential temperature
-          t2m(i)  = tskin(i)*wrk + t1(i)*prslki(i)*fhi - (grav+grav)/cp
-        else ! Use potential temperature referenced to 1000 hPa
-          t2m(i)  = tskin(i)*wrk + t1(i)*fhi - (grav+grav)/cp
-        endif
+        qv1    = q1c / (1. - q1c) ! lev 1 mixing ratio
+        qsfcmr   = qsurf(i)/(1. - qsurf(i)) ! surface mixing ratio
+        chs = cdq(i) * wind(i)
+        cqs = chs
+        qsfcprox = max(qmin,qv1 + evap(i)/cqs) ! surface mix. ratio computed from the flux
 
-        if(evap(i) >= 0.) then !  for evaporation>0, use inferred qsurf to deduce q2m
-          q2m(i) = qsurf(i)*wrk + max(qmin,q1(i))*fhi
-        else                   !  for dew formation, use saturated q at tskin
-          qss    = fpvs(tskin(i))
-          qss    = eps * qss / (ps(i) + epsm1 * qss)
-          q2m(i) = qss*wrk + max(qmin,q1(i))*fhi
-        endif
+        if(.not. flux) then
+        !-- original method
+          if(lsm /= lsm_ruc) then
+            if(thsfc_loc) then ! Use local potential temperature
+              t2m(i)=tskin(i)*wrk + t1(i)*prslki(i)*fhi - (grav+grav)/cp
+            else ! Use potential temperature referenced to 1000 hPa
+              t2m(i) = tskin(i)*wrk + t1(i)*fhi - (grav+grav)/cp
+            endif
+            if(evap(i) >= 0.) then !  for evaporation>0, use inferred qsurf to deduce q2m
+              q2m(i) = qsurf(i)*wrk + max(qmin,q1c)*fhi
+            else                   !  for dew formation, use saturated q at tskin
+              qss    = fpvs(tskin(i))
+              qss    = eps * qss/(ps(i) + epsm1 * qss)
+              q2m(i) = qss*wrk + max(qmin,q1c)*fhi
+            endif
+          else
+          !-- RUC lsm
+            if(chs.lt.1.E-5) then
+            !-- under very stable conditions use first level temperature
+              t2m(i) = t1(i)
```

Comment:
I can only approve this PR if we never use this "use first model level" hack. I'd prefer it if this option was deleted.

Reply:
@joeolson42 Joe, This is exactly what we have in HRRR 2-m diagnostics. We should discuss if we want to remove this option in RRFS at the RRFS meeting.

Reply:
@joeolson42 I removed the lines in question and repeated my test - the results were identical. Thus, it is a very rare condition, and most likely does not have any noticeable impact on 2-m diagnostics. Plus, it is used only with RUC LSM. I hope we could merge this PR as soon as possible as it contains some urgent changes for RRFS.

Reply:
I'm glad to hear the condition where those lines are applied is rare, but I will only be satisfied if they are never applied. Plus, I don't think hacks like that will make it into the community repository, so they should probably be cut out now. Can you update the PR with those lines removed?

Reply:
@joeolson42 @SamuelTrahanNOAA Joe, you are strongly opposing the algorithm which is used in HRRR and also which is available to the WRF community. Also, the 2-m diagnostics in this PR is designed to be used **only** in the FV3_HRRR suite. I do not expect any objections from the community, when there is no impact on their work. In my CCPP work I haven't experienced any objections to the code that has impact only on RUC LSM. At this point any other change will need another regression test and will delay other important RRFS_dev changes in this PR. 
@SamuelTrahanNOAA I hope this PR could go to RRFS_dev as soon as possible.

Reply:
It was always a bad idea. Using it in HRRRv4 doesn't justify it. We've often seen a warm bias in 2-m temperatures at night and this hack can only contribute to it.  Why cling to bad ideas?

Reply:
Obviously it is not perfect, but I can't agree that it is a bad idea. When there is no mixing in the surface layer, the atm. temperature near the surface is changing very little. Thus, for these conditions it is more likely to get a better estimate of 2-m temperature by simply setting it to the 1st level (8-m) temperature, than by computing it from the gradient between the skin temperature and 8-m temperature. To me, the former idea is better than the latter.

Reply:
But this hack will always result in a high estimate. Anyone who knows stable boundary layer knows that stratification continues beyond the first 2 meters of the atmosphere. In many cases, with stagnant winds, the temperature difference between the surface and ~ 10 m (approx height of the first model level) can easily exceed 10 C. Even if 80% of the temperature difference is concentrated below 2 m, that still leaves a non-negligible difference between 2 m and the first model level. Using this hack, guarantees a high estimate. It's more than imperfect, it's misleading -- a blatant attempt to hide the old nighttime cold-bias problem. I'm sorry, but his hack should go. Our code should be better than this.

Reply:
You make the commit and I'll help run the reg tests.